File: C++/TwoPointer/merge.cpp
/*
Solution:双指针，时间复杂度O(n),空间复杂度O(1)
- nums1长度足够大，把nums1中的数往后移，空出n个位置来；
- 定义两个指针p,q,分别指向nums1和nums2中的数值，比较大小即可；
*/
class Solution {
public:
	void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {

		if (n != 0) {
			for (int i = m + n - 1; i >= n; i--) {
				nums1[i] = nums1[i-n];
			}
			int p = 0, q = n, i = 0;
			while (p < n && q < (m + n)) {
				if (nums2[p] < nums1[q]) {
					nums1[i++] = nums2[p++];
				}
				else {
					nums1[i++] = nums1[q++];
				}
			}
			while (p < n) {
				nums1[i++] = nums2[p++];
			}
			while (q < (m + n)) {
				nums1[i++] = nums1[q++];
			}
		}
	}
};

File: Python3/Array/merge.py
# -*- coding:UTF-8 -*-
'''
Solution:
- 双指针
- 从两个数组的后面比较，比较大的放到nums1的末尾；
- 如果从小的开始比较，nums1里面有比较多的数字移动，时间复杂度比较高；
'''
class Solution:
    def merge(self, nums1: 'List[int]', m: 'int', nums2: 'List[int]', n: 'int') -> 'None':
        """
        Do not return anything, modify nums1 in-place instead.
        """
        i, j = m - 1, n - 1
        p = len(nums1) - 1 
        while i >=0 and j >= 0:
            if nums2[j] > nums1[i]:
                nums1[p] = nums2[j]
                j -= 1
            else:
                nums1[p] = nums1[i]
                i -= 1
            p -= 1
            
        while j >= 0:
            nums1[p] = nums2[j]
            j, p = j - 1, p - 1
